plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java

```diff
@@ -0,0 +1,335 @@
+                                                            }, p -> p.getValue()));
+
+            // add to this map only port without a known ref name
+            Map<String, String> portsToUnkownRefName =
+                    exposedPorts.stream().filter((port) -> !portsToKnownRefName.containsKey(port))
+                                .collect(Collectors.toMap(p -> p, p -> "Server-" + p.replace('/', '-')));
+
+            // list of all ports with refName (known/unknown)
+            this.portsToRefName = new HashMap(portsToKnownRefName);
+            portsToRefName.putAll(portsToUnkownRefName);
+        }
+
+        /**
+         * Gets default external address.
+         */
+        protected String getExternalAddress() {
+            return externalAddressProperty != null ?
+                   externalAddressProperty : internalAddressProperty;
+        }
+
+        /**
+         * Populate the template properties
+         */
+        protected void populateGlobalProperties() {
+            String externalAddress = getExternalAddress();
+            String externalIP = getExternalIp(externalAddress);
+            globalPropertiesMap.put("internalIp", internalAddressProperty);
+            globalPropertiesMap.put("externalIp", externalAddressProperty);
+            globalPropertiesMap.put("externalAddress", externalAddress);
+            globalPropertiesMap.put("externalIP", externalIP);
+            globalPropertiesMap.put("workspaceId", getWorkspaceId());
+            globalPropertiesMap.put("machineName", getMachineName());
+            globalPropertiesMap.put("wildcardNipDomain", getWildcardNipDomain(externalAddress));
+            globalPropertiesMap.put("wildcardXipDomain", getWildcardXipDomain(externalAddress));
+            globalPropertiesMap.put("chePort", chePort);
+        }
+
+        /**
+         * Rendering
+         */
+        public String render(String template, String port) {
+            ST stringTemplate = new ST(template);
+            globalPropertiesMap.forEach((key, value) -> stringTemplate.add(key, value));
+            stringTemplate.add("serverName", portsToRefName.get(port));
+            return stringTemplate.render();
+        }
+
+        /**
+         * Gets the workspace ID from the config of the given container
+         *
+         * @return workspace ID
+         */
+        protected String getWorkspaceId() {
+            return Arrays.stream(env).filter(env -> env.startsWith(CHE_WORKSPACE_ID_PROPERTY))
+                         .map(s -> s.substring(CHE_WORKSPACE_ID_PROPERTY.length()))
+                         .findFirst().get();
+        }
+
+        /**
+         * Gets the workspace Machine Name from the config of the given container
+         *
+         * @return machine name of the workspace
+         */
+        protected String getMachineName() {
+            return Arrays.stream(env).filter(env -> env.startsWith(CHE_MACHINE_NAME_PROPERTY))
```

Comment:
For some properties I agree labels is better
But I may think it's nice to have machine Name as env variable
For example, I may use the machine name in my prompt terminal instead of the default `$`character